### PR TITLE
Make channel shutdown safe

### PR DESCRIPTION
**Pull request by JonathanHenson**
…se, but there's times where it has to be), we need to hold a reference to it when we schedule the shutdown task.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by graebm** [comment]
Add a regression test for this

**Comment by graebm** [comment]
A regression test for this could be to have a handler that calls channel_shutdown() from each of its own handler->shutdown() and handler->delete() functions.

**Comment by JonathanHenson** [comment]
That wouldn’t help In this case. Channel tasks are canceled at shutdown. In this case, the channel is already shutdown and would never be canceled.

File: tests/channel_test.c [hunk 1]
```diff
@@ -715,6 +715,48 @@ static int s_test_channel_cancels_pending_tasks(struct aws_allocator *allocator,
...
+    };
+
+    ASSERT_SUCCESS(aws_mutex_lock(&test_args.mutex));
+    channel = aws_channel_new(allocator, event_loop, &callbacks);
+    ASSERT_NOT_NULL(channel);
+    ASSERT_SUCCESS(aws_condition_variable_wait(&test_args.condition_variable, &test_args.mutex));
+    ASSERT_INT_EQUALS(0, test_args.error_code);
+
+    ASSERT_SUCCESS(aws_channel_shutdown(channel, AWS_ERROR_SUCCESS));
+    ASSERT_SUCCESS(aws_condition_variable_wait(&test_args.condition_variable, &test_args.mutex));
```
Comment (justinboswell) [inline tests/channel_test.c]: Shouldn't this wait for test_args.shutdown_completed? Or at least validate it?

**Pull request merged**